T7896: Add frr profile selection

bl0way opened the pull request:
<!-- All PR should follow this template to allow a clean and transparent review -->
<!-- Text placed between these delimiters is considered a comment and is not rendered -->

## Change summary
<!--- Provide a general summary of your changes in the Title above -->
Add the possibility to change FRR profiles between `traditional` and `datacenter` accordingly to the official FRR [documentation](https://docs.frrouting.org/en/latest/basic.html#profiles)

## Types of changes
<!---
What types of changes does your code introduce? Put an 'x' in all the boxes that apply.
NOTE: Markdown requires no leading or trailing whitespace inside the [ ] for checking
the box, please use [x]
-->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)
<!-- optional: Link to related other tasks on Phabricator. -->
<!-- * https://vyos.dev/Txxxx -->
https://vyos.dev/T7896

## Related PR(s)
<!-- Link here any PRs in other repositories that are required by this PR -->

## How to test / Smoketest result
<!---
Please describe in detail how you tested your changes. Include details of your testing
environment, and the tests you ran. When pasting configs, logs, shell output, backtraces,
and other large chunks of text, surround this text with triple backtics
```
like this
```

Or provide the output of the smoketest

```
$ /usr/libexec/vyos/tests/smoke/cli/test_xxx_feature.py
test_01_simple_options (__main__.TestFeature.test_01_simple_options) ... ok
```
-->

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- The entire development process is outlined here: https://docs.vyos.io/en/latest/contributing/development.html -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [ ] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [x] My change requires a change to the documentation (https://docs.vyos.io/en/latest/configuration/system/frr.html)
- [ ] I have updated the documentation accordingly

bl0way committed:
frr: T7896: Configure frr profile with 'system frr profile' command

Co-authored-by: Christian Breunig <christian@breunig.cc>

bl0way commented:
I have read the CLA Document and I hereby sign the CLA

c-po commented on interface-definitions/system_frr.xml.in:
```diff
@@ -35,6 +35,22 @@
+              <valueHelp>
+                <format>datacenter</format>
+                <description>Reflects a single administrative domain with intradomain links using aggressive timers</description>
+              </valueHelp>
```
A `regex` constraint is missing here. Without the constraint one can fantasize any profile.

```
cpo@LR1.wue3# set system frr profile trad-foo-datacenter
[edit]
```

```suggestion
              </valueHelp>
              <constraint>
                <regex>(datacenter|traditional)</regex>
              </constraint>
```

c-po commented on interface-definitions/system_frr.xml.in:
```diff
@@ -35,6 +35,22 @@
           </leafNode>
+          <leafNode name="profile">
+            <properties>
+              <help>FRR profile</help>
```
```suggestion
              <help>Configuration profiles to adapt different defaults</help>
```

c-po commented on interface-definitions/system_frr.xml.in:
```diff
@@ -35,6 +35,22 @@
+                <description>Reflects a single administrative domain with intradomain links using aggressive timers</description>
+              </valueHelp>
+            </properties>
+          </leafNode>
```
You should make use of out `defaultValue` system here

```suggestion
            <defaultValue>traditional</defaultValue>
          </leafNode>
```

c-po commented on data/templates/frr/daemons.frr.tmpl:
```diff
@@ -105,7 +105,7 @@ valgrind_enable=no
 #watchfrr_options=""
 
-frr_profile="traditional"
+frr_profile="{{ profile if profile is vyos_defined else 'traditional' }}"
```
`else` branch not needed if `<defaultValue>` is specified in XML.

```suggestion
frr_profile="{{ profile }}"
```

sever-sever reviewed (approved):
Add ability to set the datacenter profile to FRR

bl0way commented:
Unfortunately I will need to modify a bit more deeply the code. If the profile is set in `daemons` file, the statement in `frr.conf` is taking precedence and this settings is automatically generated by FRR by defaults. Therefore it needs to be modified in `frr.conf` file too.

c-po commented:
@bl0way you can - of course - also build the `vyos-1x` package and install it to your development router using `dpkg`.

The execute the smoketests locally as described in https://docs.vyos.io/en/latest/contributing/testing.html#smoketests so in your case

```
vyos@vyos:~$ /usr/libexec/vyos/tests/smoke/cli/test_system_frr.py
```

 if you are unaware of this feature

c-po commented on python/vyos/frrender.py:
```diff
@@ -234,13 +234,15 @@ def dict_helper_nhrp_defaults(nhrp):
-                frr_defaults_cli_path, key_mangling=('-', '_')
-            )['profile']
-        }
+    frr_system_cli_path = ['system', 'frr']
```
Can you please follow the existing code style:

<img width="1057" height="133" alt="image" src="https://github.com/user-attachments/assets/8852f712-3f2b-4d53-91e2-fe82cd38cc94" />

c-po commented on interface-definitions/system_frr.xml.in:
```diff
@@ -35,6 +35,26 @@
           </leafNode>
+          <leafNode name="profile">
+            <properties>
+              <help>Configuration profiles to adapt different defaults</help>
```
```suggestion
              <help>Select configuration profile to adapt different defaults</help>
```

c-po commented on interface-definitions/system_frr.xml.in:
```diff
@@ -35,6 +35,26 @@
+              </completionHelp>
+              <valueHelp>
+                <format>traditional</format>
+                <description>Reflects defaults adhering mostly to IETF standards or common practices in wide-area internet routing</description>
```
```suggestion
                <description>Adhere mostly to IETF standards or common practices in wide-area internet routing</description>
```

c-po commented on interface-definitions/system_frr.xml.in:
```diff
@@ -35,6 +35,26 @@
+              </valueHelp>
+              <valueHelp>
+                <format>datacenter</format>
+                <description>Reflects a single administrative domain with intradomain links using aggressive timers</description>
```
```suggestion
                <description>Single administrative domain using aggressive timers</description>
```

c-po commented on python/vyos/frrender.py:
```diff
@@ -233,6 +233,17 @@ def dict_helper_nhrp_defaults(nhrp):
+            get_first_key=True,
+            with_recursive_defaults=True,
+        )['profile']
+    }
```
Why so complicated in addressing an inner key of the dict? Just use the already existing patterns - this makes the life a lot easier for the maintainers if any change adhere to already existing design / coding rules.

```suggestion
    dict['system_frr'] = conf.get_config_dict(frr_system_cli_path , key_mangling=('-', '_'),
                                              get_first_key=True, with_recursive_defaults=True)    
```

c-po commented on python/vyos/frrender.py:
```diff
@@ -733,6 +744,10 @@ def inline_helper(config_dict) -> str:
 
+        # FRR profile configuration
+        if 'frr' in config_dict and 'defaults' in config_dict['frr']:
+            output += f'frr defaults {config_dict["frr"]["defaults"]}\n'
```
```suggestion
        tmp = dict_search('system_frr.profile', config_dict):
        if tmp: output += f'frr defaults {tmp}\n'
```
```

c-po commented on src/conf_mode/system_frr.py:
```diff
@@ -19,18 +19,21 @@
 
 # path to daemons config and config status files
-config_file = '/etc/frr/daemons'
+deamon_config_file = '/etc/frr/daemons'
```
It's spelled `daemon` not `deamon`
```suggestion
daemon_config_file = '/etc/frr/daemons'
```

c-po commented on src/conf_mode/system_frr.py:
```diff
@@ -39,9 +42,15 @@ def get_config(config=None):
-    frr_config = conf.get_config_dict(base, key_mangling=('-', '_'),
-                                      get_first_key=True,
-                                      with_recursive_defaults=True)
+    frr_config = {
```
Why reformat the code and now nesting things inside a dict? We do not use such a coding pattern anywhere - it makes maintainer life harder. Keep the current coding style.

c-po commented on src/conf_mode/system_frr.py:
```diff
@@ -51,15 +60,25 @@ def verify(frr_config):
+    daemons_config_current = read_file(deamon_config_file)
     # generate new config file
-    daemons_config_new = render_to_string('frr/daemons.frr.tmpl', frr_config)
+    daemons_config_new = render_to_string('frr/daemons.frr.tmpl', frr_config['daemons'])
```
Unnecessary complication

c-po commented:
Please also squash all existing commits. Update commit message with CLI example how to sue this addition.

bl0way commented:
I implemented your suggestions and fixed the failing test (tested locally), so it should be all clear now, I hope.

I apologize for the spam that this PR created, and I appreciate the patience that you are taking to review and guide me through this PR. I understand better the internal logic, and I hope I will be able to contribute more in the future.

sever-sever reviewed (approved):
Add ability to configure the `datacenter` profile
```
vyos@r14# set system frr profile datacenter 
[edit]
vyos@r14# commit
[ system frr ]

WARNING: You need to reboot the router (preferred) or restart FRR to
apply changes in modules settings


[edit]
vyos@r14# run restart all
WARNING: This is a potentially unsafe function!
You may lose the connection to the router or active configuration after
running this command. Use it at your own risk!

Continue? [y/N] y
[6928|mgmtd] sending configuration
[6929|zebra] sending configuration
[6930|ripd] sending configuration
[6931|ripngd] sending configuration
[6929|zebra] done
[6930|ripd] done
[6935|bgpd] sending configuration
[6933|ospf6d] sending configuration
[6936|isisd] sending configuration
[6934|ldpd] sending configuration
[6932|ospfd] sending configuration
[6938|nhrpd] sending configuration
[6940|babeld] sending configuration
[6931|ripngd] done
[6932|ospfd] done
[6934|ldpd] done
[6928|mgmtd] done
[6933|ospf6d] done
[6935|bgpd] done
[6936|isisd] done
Waiting for children to finish applying config...
[6938|nhrpd] done
[6949|pim6d] sending configuration
[6942|fabricd] sending configuration
[6943|watchfrr] sending configuration
[6945|staticd] sending configuration
[6940|babeld] done
[6946|bfdd] sending configuration
[6949|pim6d] done
[6942|fabricd] done
[6943|watchfrr] done
[6945|staticd] done
[6946|bfdd] done
[edit]
vyos@r14# 
[edit]
vyos@r14# ps ax | grep frr
   1375 ?        S<s    0:11 /usr/lib/frr/watchfrr -d -F traditional zebra mgmtd bgpd ripd ripngd ospfd ospf6d isisd babeld pim6d ldpd nhrpd staticd bfdd fabricd
   6869 ?        Ssl    0:00 /usr/lib/frr/zebra -d -F datacenter --daemon -A 127.0.0.1 -s 90000000
   6874 ?        Ss     0:00 /usr/lib/frr/mgmtd -d -F datacenter --daemon -A 127.0.0.1
   6876 ?        Ssl    0:00 /usr/lib/frr/bgpd -d -F datacenter --daemon -A 127.0.0.1 -M rpki
   6883 ?        Ss     0:00 /usr/lib/frr/ripd -d -F datacenter --daemon -A 127.0.0.1
   6886 ?        Ss     0:00 /usr/lib/frr/ripngd -d -F datacenter --daemon -A ::1
   6889 ?        Ss     0:00 /usr/lib/frr/ospfd -d -F datacenter --daemon -A 127.0.0.1
   6892 ?        Ss     0:00 /usr/lib/frr/ospf6d -d -F datacenter --daemon -A ::1
   6895 ?        Ss     0:00 /usr/lib/frr/isisd -d -F datacenter --daemon -A 127.0.0.1
   6898 ?        Ss     0:00 /usr/lib/frr/babeld -d -F datacenter --daemon -A 127.0.0.1
   6901 ?        Ss     0:00 /usr/lib/frr/pim6d -d -F datacenter --daemon -A ::1
   6906 ?        S      0:00 /usr/lib/frr/ldpd -L -u frr -g frr
   6907 ?        S      0:00 /usr/lib/frr/ldpd -E -u frr -g frr
   6909 ?        Ss     0:00 /usr/lib/frr/ldpd -d -F datacenter --daemon -A 127.0.0.1
   6913 ?        Ss     0:00 /usr/lib/frr/nhrpd -d -F datacenter --daemon -A 127.0.0.1
   6917 ?        Ss     0:00 /usr/lib/frr/staticd -d -F datacenter --daemon -A 127.0.0.1
   6922 ?        Ss     0:00 /usr/lib/frr/bfdd -d -F datacenter --daemon -A 127.0.0.1
   6925 ?        Ss     0:00 /usr/lib/frr/fabricd -d -F datacenter --daemon -A 127.0.0.1
   6976 pts/0    S+     0:00 grep frr
[edit]
vyos@r14# 
[edit]
vyos@r14# cat /etc/fr
freeipmi/ frr/      
[edit]
vyos@r14# cat /etc/frr/daemons | grep profil
frr_profile="datacenter"
[edit]
vyos@r14# 
```
By default `traditional`
```
vyos@r14# delete system frr 
[edit]
vyos@r14# commit
[ system frr ]

WARNING: You need to reboot the router (preferred) or restart FRR to
apply changes in modules settings


[edit]
vyos@r14# cat /etc/frr/daemons | grep profil
frr_profile="traditional"
[edit]
vyos@r14# 

```

c-po commented on src/conf_mode/system_frr.py:
```diff
@@ -51,15 +56,27 @@ def verify(frr_config):
 
+    # profile could be automatically generated by frr in frr.conf
+    # and needs to be updated as it is taking precedence
+    if 'frrender' in frr_config and 'profile' in frr_config['frrender']:
```
SHould be splitted into generate and apply - see https://github.com/vyos/vyos-1x/blob/current/src/conf_mode/interfaces_ethernet.py#L358-L365

c-po committed:
frr: T7896: use proper XML CLI defaults in profile testcase

c-po reviewed (approved):
Changing profile and restarting FRR restarts daemons with new profile.

Deleting profile from CLI restores previous `tradition` profile which is the default.
